plugins/InspectionGadgets/InspectionGadgetsAnalysis/src/com/siyeh/ig/internationalization/UnnecessaryUnicodeEscapeInspection.java

```diff
@@ -113,69 +175,24 @@ public void visitFile(PsiFile file) {
-      final CharBuffer charBuffer = CharBuffer.allocate(1);
-      final ByteBuffer byteBuffer = ByteBuffer.allocate(10);
-      final int length = text.length();
-      for (int i = 0; i < length; i++) {
-        final char c = text.charAt(i);
-        if (c != '\\') {
-          continue;
-        }
-        boolean isEscape = true;
-        int previousChar = i - 1;
-        while (previousChar >= 0 && text.charAt(previousChar) == '\\') {
-          isEscape = !isEscape;
-          previousChar--;
-        }
-        if (!isEscape) {
-          continue;
-        }
-        int nextChar = i;
-        do {
-          nextChar++;
-          if (nextChar >= length) {
-            break;
-          }
-        }
-        while (text.charAt(nextChar) == 'u'); // \\uuuu0061 is a legal unicode escape
-        if (nextChar == i + 1 || nextChar + 3 >= length) {
-          continue;
-        }
-        if (StringUtil.isHexDigit(text.charAt(nextChar)) &&
-            StringUtil.isHexDigit(text.charAt(nextChar + 1)) &&
-            StringUtil.isHexDigit(text.charAt(nextChar + 2)) &&
-            StringUtil.isHexDigit(text.charAt(nextChar + 3))) {
-          final int escapeEnd = nextChar + 4;
-          final char d = (char)Integer.parseInt(text.substring(nextChar, escapeEnd), 16);
-          final int type = Character.getType(d);
-          if (type == Character.CONTROL ||
-              type == Character.FORMAT ||
-              type == Character.PRIVATE_USE ||
-              type == Character.SURROGATE ||
-              type == Character.UNASSIGNED ||
-              type == Character.LINE_SEPARATOR ||
-              type == Character.PARAGRAPH_SEPARATOR) {
-            continue;
-          }
-          if (type == Character.SPACE_SEPARATOR && d != ' ') {
-            continue;
-          }
-          byteBuffer.clear();
-          charBuffer.clear();
-          charBuffer.put(d).rewind();
-          final CoderResult coderResult = encoder.encode(charBuffer, byteBuffer, true);
-          if (coderResult.isError()) {
-            continue;
-          }
-          final PsiElement element = file.findElementAt(i);
+
+      forEachUnicodeSequence(text, charset, new Callback() {
+        @Override
+        public boolean foundSequence(String text, int start, int end, char codeUnit) {
+          final PsiElement element = file.findElementAt(start);
           if (element != null && isSuppressedFor(element)) {
-            return;
+            return false;
           }
-          final RangeMarker rangeMarker = document.createRangeMarker(i, escapeEnd);
-          registerErrorAtOffset(file, i, escapeEnd - i, Character.valueOf(d), rangeMarker);
+          final RangeMarker rangeMarker = document.createRangeMarker(start, end);
+          registerErrorAtOffset(file, start, end - start, Character.valueOf(codeUnit), rangeMarker);
+          return true;
         }
-      }
+      });
     }
   }
+
+  // @VisibleForTesting
```

Comment:
Did you forget some commented annotation?

Reply:
No, I didn't. It looks like an annotation, but since IDEA doesn't have this annotation imported by default, I cannot remove the comment. Still, I find it nice to have as kind of documentation.